core/src/utils/animation/animation.ts

```diff
@@ -0,0 +1,816 @@
+
+  const afterAddRead = (readFn: () => void): Animation => {
+    _afterAddReadFunctions.push(readFn);
+
+    return generatePublicAPI();
+  };
+
+  const afterAddWrite = (writeFn: () => void): Animation => {
+    _afterAddWriteFunctions.push(writeFn);
+
+    return generatePublicAPI();
+  };
+
+  /**
+   * Add CSS class to this animation's elements
+   * before the animation begins.
+   */
+  const beforeAddClass = (className: string | string[] | undefined): Animation => {
+    beforeAddClasses = addClassToArray(beforeAddClasses, className);
+
+    return generatePublicAPI();
+  };
+
+  /**
+   * Remove CSS class from this animation's elements
+   * before the animation begins.
+   */
+  const beforeRemoveClass = (className: string | string[] | undefined): Animation => {
+    beforeRemoveClasses = addClassToArray(beforeRemoveClasses, className);
+
+    return generatePublicAPI();
+  };
+
+  /**
+   * Set CSS inline styles to this animation's elements
+   * before the animation begins.
+   */
+  const beforeStyles = (styles: { [property: string]: any } = {}): Animation => {
+    beforeStylesValue = styles;
+
+    return generatePublicAPI();
+  };
+
+  /**
+   * Clear CSS inline styles from this animation's elements
+   * before the animation begins.
+   */
+  const beforeClearStyles = (propertyNames: string[] = []): Animation => {
+    for (const property of propertyNames) {
+      beforeStylesValue[property] = '';
+    }
+
+    return generatePublicAPI();
+  };
+
+  /**
+   * Add CSS class to this animation's elements
+   * after the animation ends.
+   */
+  const afterAddClass = (className: string | string[] | undefined): Animation => {
+    afterAddClasses = addClassToArray(afterAddClasses, className);
+
+    return generatePublicAPI();
+  };
+
+  /**
+   * Remove CSS class from this animation's elements
+   * after the animation ends.
+   */
+  const afterRemoveClass = (className: string | string[] | undefined): Animation => {
+    afterRemoveClasses = addClassToArray(afterRemoveClasses, className);
+
+    return generatePublicAPI();
+  };
+
+  /**
+   * Set CSS inline styles to this animation's elements
+   * after the animation ends.
+   */
+  const afterStyles = (styles: { [property: string]: any } = {}): Animation => {
+    afterStylesValue = styles;
+
+    return generatePublicAPI();
+  };
+
+  /**
+   * Clear CSS inline styles from this animation's elements
+   * after the animation ends.
+   */
+  const afterClearStyles = (propertyNames: string[] = []): Animation => {
+    for (const property of propertyNames) {
+      afterStylesValue[property] = '';
+    }
+
+    return generatePublicAPI();
+  };
+
+  const getFill = (): 'auto' | 'none' | 'forwards' | 'backwards' | 'both' | undefined => {
+    if (_fill !== undefined) { return _fill; }
+    if (parentAnimation) { return parentAnimation.getFill(); }
+
+    return undefined;
+  };
+
+  const getDirection = (): 'normal' | 'reverse' | 'alternate' | 'alternate-reverse' | undefined => {
```

Comment:
this `'normal' | 'reverse' | 'alternate' | 'alternate-reverse'` is used many places, let's create a TS type: 

KeyframeDirection? maybe?